.pre-commit-config.yaml

```diff
@@ -0,0 +1,63 @@
+# Why should I care about or use the pre-commit tool?
+#    Automate away your pre-commit and pre-push checks and steps.
+#    The longer and more nuanced answer is in the links below.
+#
+# Pre-commit Documentation: https://pre-commit.com/
+# Pre-commit hooks documentation: https://pre-commit.com/hooks.html
+#
+# Tool installation:
+#   $pip install pre-commit
+#
+# All pre-commit commands are to be run from the root dir of the repo,
+# or wherever the '.pre-commit-config.yaml' file is located.
+#
+# Hooks default (local) installation:
+#   $pre-commit install
+#
+# Hooks (local) removal:
+#   $pre-commit uninstall
+#
+# Install only a subset of the hooks:
+#   $pre-commit install --hook-type pre-commit
+#
+# Run default hooks manually:
+# This only works on staged un-committed files.
+# Unstaged files are stashed, the checks performed, then the stash is restored.
+#   $pre-commit run
+#
+# Run default hooks on all files in repo:
+#   $pre-commit run --all-files
+#
+# [For CI/DevOps only]
+# Hook repos version update (run in the root dir):
+#   $pre-commit autoupdate
+
+default_install_hook_types: [pre-commit, pre-push]
+
+default_stages: [pre-commit]
+
+repos:
+    # Rust hooks.
+    - repo: https://github.com/doublify/pre-commit-rust
+      rev: v1.0
+      hooks:
+        # FIXME: enable fmt hook when it no longer emits warnings with
+        # stable build. Currently it requires nightly build.
+        # Until then, devs should run `cargo +nightly fmt` manually.
+        #- id: fmt
+        - id: clippy
+          stages: [pre-push]
+        - id: cargo-check
+          stages: [pre-push]
```

Comment:
We probably do not need to cater to everyone, but a lot of people create many commits that they expect GitHub to squash for them [^1].
I am a bit concerned that these will be too slow.

Our CI uses the following scripts:
1. `fmt`: `scripts/cargo-fmt.sh`
2. `clippy`: `scripts/cargo-clippy.sh`
3. I'm actually not sure the exact command to run all the tests.

But even the `fmt` script is quite slow.
If I need to run `fmt` myself, I always specify the affected package explicitly, as in:
```bash
./cargo nightly fmt --package solana-core
```

`clippy` is even slower.

And if you run the wrong version, you might get discrepancies from what you get later in CI.

[^1]: I personally prefer a different style of creating changes.
But, unfortunately, GitHub review process is more or less designed to work this way :(